uaclient/cli/commands.py

```diff
@@ -0,0 +1,50 @@
+import argparse
+from typing import Callable, List, Optional
+
+from uaclient.cli.constants import NAME, USAGE_TMPL
+from uaclient.config import UAConfig
+
+
+class ProArgument:
+    def __init__(
+        self, long_name: str, help: str, short_name: Optional[str] = None
+    ):
+        self.names = (
+            (long_name,) if short_name is None else (short_name, long_name)
+        )
+        self.help = help
+
+    def register(self, parser: argparse.ArgumentParser):
+        parser.add_argument(*self.names, help=self.help)
+
+
+class ProCommand:
+    def __init__(
+        self,
+        name: str,
+        help: str,
+        description: str,
+        usage: Optional[str] = None,
+        action: Callable = lambda *args: None,
+        arguments: List[ProArgument] = [],
+    ):
+        self.name = name
+        self.help = help
+        self.description = description
+        self.usage = usage or USAGE_TMPL.format(name=NAME, command=name)
+        self.action = action
+        self.arguments = arguments
+
+    def __call__(self, args: argparse.Namespace, *, cfg: UAConfig, **kwargs):
+        return self.action(args, cfg)
```

Comment:
Defining `__call__` and passing `self` as the action feels a bit weird to me.

1. Do we need this layer of indirection? What is the benefit of this over setting `action=self.action` during `register`?
2. If we do need this layer - could we accomplish the same via defining a method here instead of `__call__`? It could be something like:
```suggestion
    def call_action(self, args: argparse.Namespace, *, cfg: UAConfig, **kwargs):
        return self.action(args, cfg)
```
 and pass `action=self.call_action` in `register`.

Reply:
![obrazek](https://github.com/canonical/ubuntu-pro-client/assets/15742644/4b526ca4-832a-46e7-b948-2fa5e563ae77)

Reply:
Jokes apart we can change it to not use the dunders, sure we do

I mean I would use more and more dunder methods instead of helpers, here and in the code overall.
I took this one particularly from a suggestion in SO discussions on how to make argparse more OOPythinic and found it to be interesting

Reply:
:laughing: pythonicity is in the eye of the beholder I guess? :upside_down_face: 
If the extra layer is useful, I don't mind using the dunder all that much. But is it useful? It would only be useful if we wanted to do something additional in the `__call__` method besides drill the args into `self.action`? If we foresee that being something we may want to do in the future, then I'm okay with it too.

Reply:
Right nice, I'm up to that. For now I will convert it to a regular method and pass it. If something happens, won't be a huge change.

Reply:
the `.action` in the tests now...
hahahaha
resolving this one.